## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
column_list = list(df.columns)
column_list


['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [6]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts()
top_categories.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries, which subcategory had the most entries?
top_subcategory = df['subcategory'].value_counts()
top_subcategory.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts()
top_clients.head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
top_client_list_result = top_clients.sort_values(ascending=False).head()
top_client_list = top_client_list_result.index.tolist()
top_client_list

[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_df = df[df['client_id'] == 33615]
top_client_qty = top_client_df['qty'].sum()
top_client_qty

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
subtotal = df["unit_price"]*df['qty']
df["line_subtotal"] = subtotal
df[["unit_price", "qty", "line_subtotal"]].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
total_weight = df["unit_weight"]*df['qty']
df["total_weight"] = total_weight

shipping = df.apply(lambda x: x['total_weight'] * 7 \
                    if x['total_weight'] > 50 else x['total_weight'] * 10, axis=1)

shipping_price = df["shipping_price"] = shipping



df[["unit_price", "unit_weight", "qty", "total_weight", "shipping_price"]].head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
tax = 0.0925
total_price = df['line_subtotal'] +df['shipping_price']
df["total_price"] = total_price
df["total_price"] *= 1 + tax

df[['line_subtotal', 'shipping_price', 'total_price']].head(3)

line_subtotal  shipping_price    total_price
0      115164.00         5512.50  131839.076250
1         523.95          312.90     914.258625
2         527.28          458.64    1077.117600

In [14]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
line_cost = (df['unit_cost']*df['qty']) + df['shipping_price']
df['line_cost'] = line_cost
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  total_weight  shipping_price    total_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   
3            3        1056.18         35.67          356.70    1543.571400   
4            1        3902.00        928.60         6500.20   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [15]:
# Create a column for the profit of each line using line cost and line price
line_profit = df['total_price'] - df['line_cost']
df['line_profit'] = line_profit

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  shipping_price    total_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   
3        1056.18         35.67          356.70    1543.571400    1077.35   
4        3902.00        928.60         6500.20   11364.403500    8663.60   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
order1 = df.loc[df['order_id'] == 2742071]
order1_total = order1['total_price'].sum()
order1_total
print(f"Order 2742071 total: ${order1_total:,.2f}")

order2 = df.loc[df['order_id'] == 2173913]
order2_total = order2['total_price'].sum()
print(f"Order 2173913 total: ${order2_total:,.2f}")

order3 = df.loc[df['order_id'] == 6128929]
order3_total = order3['total_price'].sum()
print(f"Order 6128929 total: ${order3_total:,.2f}")


Order 2742071 total: $152,811.90
Order 2173913 total: $162,388.72
Order 6128929 total: $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Calculate the total cost for the top 5 clients

top_clients_cost1 = df[df['client_id'].isin(top_client_list_result.index)].groupby('client_id').apply(lambda x: (x['qty'] * x['unit_cost']).sum())

# Sort the result in descending order
top_clients_cost = top_clients_cost1.sort_values(ascending=False)

# Print the formatted output
for client_id, total_cost in top_clients_cost.items():
    print(f"{client_id}: ${total_cost:,.2f}")


24741: $40,562,451.34
38378: $6,205,265.58
66037: $5,609,331.13
46820: $5,405,742.80
33615: $4,346,329.02


In [18]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5_clients_ids = df['client_id'].value_counts().head(5).index

# Calculate total units purchased by each of the top 5 clients
top_clients_units = df[df['client_id'].isin(top_5_clients_ids)].groupby('client_id')['qty'].sum()

# Calculate total shipping price from each of the top 5 clients
top_clients_shipping = df[df['client_id'].isin(top_5_clients_ids)].groupby('client_id').apply(lambda x: (x['shipping_price']).sum())

# Calculate total revenue from each of the top 5 clients
top_clients_revenue = df[df['client_id'].isin(top_5_clients_ids)].groupby('client_id').apply(lambda x: (x['qty'] * x['unit_price']).sum())

# Calculate total cost from each of the top 5 clients
top_clients_cost = df[df['client_id'].isin(top_5_clients_ids)].groupby('client_id').apply(lambda x: (x['qty'] * x['unit_cost']).sum())

# Calculate total profit from each of the top 5 clients
top_clients_profit = top_clients_revenue - top_clients_cost

# Assign names to the Series
top_clients_shipping.name = 'Total Shipping Price (millions)'
top_clients_revenue.name = 'Total Revenue (millions)'
top_clients_profit.name = 'Total Profit (millions)'

# Merge 'first' and 'last' columns with 'client_id' from the main data
top_clients_info = df[['client_id', 'first', 'last']].drop_duplicates().set_index('client_id')

# Merge with top clients' units purchased
top_clients_info = top_clients_info.merge(top_clients_units, how='inner', left_index=True, right_index=True)

# Merge with top clients' total shipping price
top_clients_info = top_clients_info.merge(top_clients_shipping, how='inner', left_index=True, right_index=True)

# Merge with top clients' total revenue
top_clients_info = top_clients_info.merge(top_clients_revenue, how='inner', left_index=True, right_index=True)

# Merge with top clients' total profit
top_clients_info = top_clients_info.merge(top_clients_profit, how='inner', left_index=True, right_index=True)

# Create the summary DataFrame
summary_df = top_clients_info.reset_index()

# Sort the summary DataFrame by total profit
summary_df = summary_df.sort_values(by='Total Profit (millions)', ascending=False)

summary_df

client_id      first     last     qty  Total Shipping Price (millions)  \
3      24741     Kendra  Garrett  239862                       5126448.37   
2      66037     Angela  Everett   43018                       1395151.85   
4      38378  Alexandra    Young   73667                       3429455.40   
1      46820      Bryan    Myers   75768                       1601448.84   
0      33615    Jessica    Reyes   64313                       1828984.89   

   Total Revenue (millions)  Total Profit (millions)  
3               70176885.25              29614433.91  
2                7995708.38               2386377.25  
4                8384321.15               2179055.57  
1                7317356.03               1911613.23  
0                5839032.11               1492703.09

In [20]:
# Define a function that converts a dollar amount to millions.
def convert_to_millions(summary_df, column_name):
    """
    Converts the values of the specified column in the DataFrame to millions.

    Parameters:
    df (pd.DataFrame): The DataFrame containing the column to be converted.
    column_name (str): The name of the column to convert.

    Returns:
    pd.DataFrame: The DataFrame with the specified column values converted to millions.
    """
    summary_df[column_name] = summary_df[column_name] / 1000000
    return summary_df
# List of columns to convert to millions # Apply the currency_format_millions function to only the money columns. 
columns_to_convert = ['Total Shipping Price (millions)', 'Total Revenue (millions)', 'Total Profit (millions)']
# Convert the specified columns to millions # Apply the currency_format_millions function to only the money columns. 
summary_df_millions = convert_to_millions(summary_df, columns_to_convert)
# Convert the specified columns to millions
summary_df_millions




client_id      first     last     qty  Total Shipping Price (millions)  \
3      24741     Kendra  Garrett  239862                         5.126448   
2      66037     Angela  Everett   43018                         1.395152   
4      38378  Alexandra    Young   73667                         3.429455   
1      46820      Bryan    Myers   75768                         1.601449   
0      33615    Jessica    Reyes   64313                         1.828985   

   Total Revenue (millions)  Total Profit (millions)  
3                 70.176885                29.614434  
2                  7.995708                 2.386377  
4                  8.384321                 2.179056  
1                  7.317356                 1.911613  
0                  5.839032                 1.492703

In [22]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
Profit_sort = summary_df_millions.sort_values(by='Total Profit (millions)', ascending=False)
Profit_sort

client_id      first     last     qty  Total Shipping Price (millions)  \
3      24741     Kendra  Garrett  239862                         5.126448   
2      66037     Angela  Everett   43018                         1.395152   
4      38378  Alexandra    Young   73667                         3.429455   
1      46820      Bryan    Myers   75768                         1.601449   
0      33615    Jessica    Reyes   64313                         1.828985   

   Total Revenue (millions)  Total Profit (millions)  
3                 70.176885                29.614434  
2                  7.995708                 2.386377  
4                  8.384321                 2.179056  
1                  7.317356                 1.911613  
0                  5.839032                 1.492703

## Part 4: Summary

Our top client, Kendra Garrett, purchased 239,862 units, generating $70,176,885.25 in revenue and $29,614,433.91 in profit. Angela Everett follows with 43,018 units and $2,386,377.25 in profit. Alexandra Young ranks third with 73,667 units and $2,179,055.57 in profit. Bryan Myers is fourth with 75,768 units and $1,911,613.23 in profit. Lastly, Jessica Reyes bought 64,313 units, earning $1,492,703.09 in profit.